apps/noodle/librapi2/branches/0.9.3-3guada1/tools/man/rapiconfig.1
.\" $Id: rapiconfig.1 2348 2006-04-05 14:58:44Z twogood $
.TH "RAPICONFIG" "1" "August 2003" "The SynCE project" "http://synce.sourceforge.net/"
.SH NAME
rapiconfig \- upload an XML config file to a SynCE connected device

.SH SYNOPSIS
\fBrapiconfig\fR [-d \fILEVEL\fR] [-P DEVPATH] [-h] [-p] [-m] \fIFILENAME\fR

.SH "DESCRIPTION"

.PP
\fBrapiconfig\fR uploads an XML config file to a device connected via SynCE.

.SH "OPTIONS"
.TP
-d \fILEVEL\fR
Set debug log level:
.IP
0 - No logging (default)
.IP
1 - Errors only
.IP
2 - Errors and warnings
.IP
3 - Everything

.TP
-p \fIDEVPATH\fR
Use the device at the given path, instead of the default.


.TP
-h
Display help message.

.TP
-p
Process the document.

.TP
-m
Return metadata.

.TP
\fIFILENAME\fR
Name of the XML configuration document.

.SH "AUTHOR"
.PP
David Eriksson <twogood@users.sourceforge.net>.
.SH "SEE ALSO"
synce(1)


Changelog
BETA3 (n/a)
===========

PERSONALIZACIÓN Y CONFIGURACIÓN DEL SISTEMA (BRANDING)
------------------------------------------------------

- Cerrada lista prácticamente definitiva de aplicaciones
- Incluidas nuevas aplicaciones:
  - lsr
  - mozilla-acroread
  - dasher
  - gFTP (BTS 1087)
  - xchat (BTS 1086)
  - bluefish (BTS 1085)
  - d4x (BTS 1084)
  - soundtracker (BTS 1063)
  - gnome-games
  - firestarter
  - streamtuner
  - aMSN (v. 0.96)
  - gdebi
  - audacity (BTS 1059)
- Incluidas aplicaciones de consola (del spec): mc, ncftp, minicom, vim, ssh,
wget, rsync, dillo, sqlite, nmap, gsm-utils, brltty, joe
- Mejoras en el sistema de suplementos: comprobación de si el suplemento se
instala correctamente, comprobación de dependencias,...
- Adaptación de "contenido para Windows" de la v3 a la v4: pendiente de
terminar de depurar (BTS 1044)
- Solucionado problema de acceso a unidades SAMBA de la red: no se refrescaban
correctamente las unidades de red a través de nautilus
- Incluidos ficheros de ejemplo, adaptando los de la v3
- Pruebas de hibernación y gestión de energía en equipos portátiles correctas
- Añadido asistente para instalación de codecs de audio/vídeo
- Homogeneizadas entradas del menú: descripción (nombre aplicación)
- Implementado e incluido asistente para conexión a Internet
- Incluido "watermain"
- Traducido "gxine" (al final, descartada de la distribución base)
- Incluida imagen de fondo de Grub
- Personalizada página de prueba de impresión
- Configuración adecuada de "rhythmbox": inclusión de algunas radios
nacionales (tal y como se indica en el "spec")
- Configuración adecuada de "liferea": inclusión de algunos periódicos
nacionales (tal y como se indica en el "spec")
- Las aplicaciones en "Herramientas de Sistema" pasan a "Sistema ->
Administración"
- Configuración adecuada de "amule": navegador y reproductor de vídeo por
defecto (BTS 1017)


INSTALADOR LIVE (LIVE-INSTALLER)
--------------------------------

- Añadido visualizador de particiones de disco en selección de puntos de
montaje
- Mejorado (más usable y sencillo) el resumen final del instalador (BTS 955)
- Corregido problema de instalación de paquetes de soporte de inglés al final
de la instalación (ya no los instala)
- Corregido error de instalación cuando existen múltiples discos duros (BTS
965 y 1126)
- Mejorado sistema de redimensionado: permitía disminuir el tamaño por debajo
del mínimo necesario para la instalación
- Mejorado opción de utilizar espacio libre: permitía usar espacio sin
particionar demasiado pequeño
- Añadido soporte para instalación del perfil de accesibilidad con el que se
arranca en live al usuario que se crea en el sistema instalado
- Mejoradas descripciones de opciones de particionado
- Corregidos errores de unidades montadas tras usar "DiskPreview"
- Establecidos valores mínimos de instalación: 3 GB (/), 500 MB (/home), 64 MB
(swap)
- Análisis de soporte de accesibilidad de "gparted": habría que hacer un "fork"
- Personalizado botón de instalar incluyendo nuevo logo

INSTALADOR NO LIVE (DI-INSTALLER)
---------------------------------

- Corregidos algunos problemas leves de traducción

ACCESIBILIDAD
-------------

- Análisis y repaso de perfiles de accesibilidad
- Validación de funcionamiento de los perfiles de accesibilidad

NOODLE
------

- Implementado e incluido actor de Hermes para gestión de PDAs
- Finalizada integración en Guadalinex: paquetes guadalinex-noodle
- Análisis de la tecnología "opensync" para sincronización a través de
"syncml" (eg: symbian); probadas diferentes versiones (0.18 (dapper), 0.19,
020svn): descartado por considerarse poco estable aún.
- Corregidos algunos errores.

SISTEMA DE GENERACIÓN (GENERATION SYSTEM)
-----------------------------------------

- Actualizados paquetes relevantes de "-updates", "-security" y "-backports"
- Mejorada la estructura de metapaquetes de Guadalinex para hacer más flexible
la desinstalación de paquetes sin afectar a las configuraciones del resto

E-ADMIN
-------

- Incluido sistema de instalación de certificado raíz de la FNMT:
  - Cuando se detecta un certificado de usuario y se va a instalar en Firefox,
  se pide permiso al usuario para la instalación del certificado raíz,
  pidiendo que acepte la licencia.
  - Hecho lo anterior, se instala el certificado raíz en el sistema
- Modificado sistema de creación de usuarios para añadir a los nuevos usuarios
al grupo "scard"
- Añadidas librerías "jss" para acceso a diferentes servicios Java de la JA
- Análisis de firma en Evolution con DNIe: descartado (no es posible firmar
con el DNIe)


BETA2 (12/01/07)
================

PERSONALICACIÓN Y CONFIGURACIÓN DEL SISTEMA (BRANDING)
------------------------------------------------------

- Incluido "ndiswrapper" para soporte de algunas tarjetas wifi
- Añadidos iconos en el escritorio: "carpeta personal" y "equipo".
- Repaso e inclusión en Guadalinex de todos los paquetes interesantes que
incluye Ubuntu Edgy y que no estaban en Guadalinex: evolution-exchange,...
- Añadidos repositorios oficiales de Guadalinex v4
- Implementado interfaz de usuario para suplementos de manera que se pueda
utilizar sin hermes
- Compartición del directorio "Compartido" para permitir a los usuarios
compartir ficheros a través de SAMBA
- Regenerados "grub" y "grub-installer" para configurar adecuadamente el
nombre de la distribución ("Ubuntu" -> "Guadalinex")
- Comprobación de detección de entradas de GRUB para otros sistemas
operativos tras instalación
- Implementación de nueva funcionalidad para gnome-system-tools: soporte de
perfiles de accesibilidad para los usuarios que se crean
- Establecido número de escritorios virtuales a 4
- Añadido "alacarte" para edición de menús
- Añadido "camorama" para uso de webcams
- Añadido "lemurae" para acceso al diccionario de la RAE
- Traducido mensaje inicial al abrir una consola
- Personalizado el icono de salida de sesión
- Deshabilitado servidor SSH por defecto

apps/noodle/librapi2/branches/0.9.3-3guada1/tools/rapiconfig.c
/* $Id: rapiconfig.c 2348 2006-04-05 14:58:44Z twogood $ */
#include <rapi.h>
#include <synce.h>
#include <synce_log.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

char* devpath = NULL;

static void show_usage(const char* name)
{
	fprintf(stderr,
			"Syntax:\n"
			"\n"
			"\t%s [-d LEVEL] [-P DEVPATH ] [-h] [-p] [-m] FILENAME"
			"\n"
			"\t-d LEVEL     Set debug log level\n"
			"\t                 0 - No logging (default)\n"
			"\t                 1 - Errors only\n"
			"\t                 2 - Errors and warnings\n"
			"\t                 3 - Everything\n"
			"\t-P DEVPATH   Device path\n"
			"\t-h           Show this help message\n"
			"\t-p           Process the document\n"
			"\t-m           Return metadata\n"
			"\tFILENAME     Name of the XML configuration document\n",
			name);
}

static bool handle_parameters(int argc, char** argv, DWORD* flags, const char** filename)
{
	int c;
	int log_level = SYNCE_LOG_LEVEL_LOWEST;

  *flags = 0;

	while ((c = getopt(argc, argv, "d:hmp:")) != -1)
	{
		switch (c)
		{
			case 'd':
				log_level = atoi(optarg);
				break;
			
                        case 'P':
                                devpath = optarg;
                                break;
			
			case 'h':
			default:
				show_usage(argv[0]);
				return false;

      case 'm':
        *flags |= CONFIG_RETURN_METADATA;
        break;

      case 'p':
        *flags |= CONFIG_PROCESS_DOCUMENT;
        break;
		}
	}
				
	synce_log_set_level(log_level);

  if ((argc - optind) < 1)
  {
 		fprintf(stderr, "%s: Filename is missing on command line\n", 
				argv[0]);
    return false;
  }

  *filename = argv[optind++];

	return true;
}

#define BUFFER_SIZE   1024

static long fsize(FILE *file)
{
  long size = -1;
  if (fseek(file, 0, SEEK_END) == 0)
  {
    size = ftell(file);
    fseek(file, 0, SEEK_SET);
  }
  return size;
}


int main(int argc, char** argv)
{
  int result = 1;
  HRESULT hr = E_UNEXPECTED;
  DWORD flags = 0;
  FILE* file = NULL;
  const char* filename = NULL;
  char* buffer = NULL;
  char* p = NULL;
  long bytes_left;
  LPWSTR config = NULL;
  LPWSTR reply = NULL;
  RapiConnection* connection = NULL;

  /*
     Initialization
   */

	if (!handle_parameters(argc, argv, &flags, &filename))
		goto exit;

  /*
     Read input file and convert to UNICODE
   */

  file = fopen(filename, "r");
  if (!file)
  {
		fprintf(stderr, "%s: Failed to open file '%s': %s\n", 
        argv[0],
				filename,
				strerror(errno));
		goto exit;
  }

  bytes_left = fsize(file);
  buffer = malloc(bytes_left);

  if (!buffer)
  {
 		fprintf(stderr, "%s: Failed to allocate %li bytes\n", 
        argv[0],
				bytes_left);
		goto exit;
  }

  p = buffer;

  while (bytes_left)
  {
    size_t bytes_read = fread(p, 1, bytes_left, file);

    if (0 == bytes_read)
    {
      fprintf(stderr, "%s: Reading stopped at offset %i: %s\n",
          argv[0],
          p - buffer,
          strerror(errno));
      goto exit;
    }
    
    p += bytes_read;
    bytes_left -= bytes_read;
  }

  config = wstr_from_current(buffer);
  if (!config)
  {
    fprintf(stderr, "%s: Failed to convert file contents to UNICODE\n",
          argv[0]);
    goto exit;
  }

  free(buffer);
  buffer = NULL;

  /* 
     Do the bossanova
   */

        if ((connection = rapi_connection_from_path(devpath)) == NULL)
        {
          fprintf(stderr, "%s: Could not find configuration at path '%s'\n", 
                  argv[0],
                  devpath?devpath:"(Default)");
          goto exit;
        }
        rapi_connection_select(connection);
	hr = CeRapiInit();

	if (FAILED(hr))
	{
		fprintf(stderr, "%s: Unable to initialize RAPI: %s\n", 
				argv[0],
				synce_strerror(hr));
		goto exit;
	}

  hr = CeProcessConfig(config, flags, &reply);

  if (SUCCEEDED(hr))
  {
    result = 0;
    buffer = wstr_to_current(reply);
    printf("%s", buffer);
    CeRapiFreeBuffer(buffer);
    buffer = NULL;
  }

exit:
  CeRapiUninit();

  if (buffer)
    free(buffer);

  wstr_free_string(config);

  return result;
}

